### PR TITLE
Tool/snap aligner paired

### PR DESCRIPTION
**Pull request by matthdsm**
## PR checklist
Add aligner (part 1) to snap-aligner

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Review by sateeshperi — COMMENTED** [review]
LGTM! Lets get another review before merge!

File: modules/snapaligner/paired/main.nf
Comment (jfy133) [inline modules/snapaligner/paired/main.nf:27]: Why not just pass $index as an directory to the module in the first place?

File: modules/snapaligner/paired/meta.yml
Comment (jfy133) [inline modules/snapaligner/paired/meta.yml:26]:
```suggestion
      pattern: "*.{fastq.gz, .fq.gz, fastq, fq, bam}"
```

**Review by jfy133 — APPROVED** [review]
Couple of minor suggestions and one question

File: modules/snapaligner/paired/main.nf
Comment (matthdsm) [reply modules/snapaligner/paired/main.nf:27]: Because I didn't know the syntax 😇  and the output from `index` are just the files

Comment (matthdsm) [reply modules/snapaligner/paired/main.nf:27]: Suggestions to improve my knowledge and the code quality are welcome!

Comment (jfy133) [reply modules/snapaligner/paired/main.nf:27]: Ok fair enough!

File: modules/snapaligner/paired/main.nf
Comment (matthdsm) [reply modules/snapaligner/paired/main.nf]: fixed

File: modules/snapaligner/paired/main.nf
Comment (jfy133) [reply modules/snapaligner/paired/main.nf:27]: No, BWA takes a diretory but uses a `find` command to find the relevant files to get the 'basename' and path for the reference genome.

Fine either way tbh

**Comment by matthdsm** [comment]
Unrelated question. I saw `bwamem2` does something like this 
```
 def read_group = meta.read_group ? "-R ${meta.read_group}" : ""
```
to add the RG. Is that something that can be done from the config file `ext.args`?  Or should I add it to command line?

**Comment by matthdsm** [comment]
It doesn't, but for the sake of completeness, you can add a custom RG through that option.
That's something we'd like to do too, hence why I'm asking 😄

**Comment by matthdsm** [comment]
Wouldn't removing it break some stuff in sarek and other pipelines using bwa/bwamem2?

**Comment by jfy133** [comment]
> Wouldn't removing it break some stuff in sarek and other pipelines using bwa/bwamem2?

@maxulysse made the mistake in the first place, so that's his problem ;).

But not necessarily no, when you update a module with `nf-core modules update <tool>` it offers you to display the diff so you can see what has changed. Given each module is installed into the pipeline with a git commit hash as a 'version', it means you can keep the old one if you want ffor as long as you want.

Lucky for him I don't have the time at the moment ;)

**Comment by maxulysse** [comment]
> > Wouldn't removing it break some stuff in sarek and other pipelines using bwa/bwamem2?
> 
> @maxulysse made the mistake in the first place, so that's his problem ;).
> 
> But not necessarily no, when you update a module with `nf-core modules update <tool>` it offers you to display the diff so you can see what has changed. Given each module is installed into the pipeline with a git commit hash as a 'version', it means you can keep the old one if you want ffor as long as you want.
> 
> Lucky for him I don't have the time at the moment ;)

I hate you soooo much <3

**Comment by jfy133** [comment]
<3

**Comment by ramprasadn** [comment]
@jfy133 @maxulysse Have a question. If the option to pass the RGs were moved to config files, would we still be able to assign read groups dynamically? So for instance, raredisease pipeline creates RGs dynamically from the sample sheet. https://github.com/nf-core/raredisease/blob/e22443cfc3770667ceff9978f457079bd3869797/subworkflows/local/check_input.nf#L36 Could this be somehow done in the configs?

**Comment by jfy133** [comment]
Yes, you can just insert into the configs;

e.g.
```
ext.args = { "-R ${meta.readgroups}" }
```

We are actually using this heavily in nf-core/taxprofiler, where actually [our entire argument/parameter](https://github.com/nf-core/taxprofiler/blob/1ac5a129a8ea3a212682e5545a8e657de16559bf/conf/modules.config#L199) set for the tool is passed via a meta element

**Comment by maxulysse** [comment]
Yes of course, you just use ext.args to print out meta.read_group

**Comment by ramprasadn** [comment]
Ah, like that. I misunderstood your point @jfy133. I thought you were talking about removing readgroup information from meta, now it all makes sense 😅

**Pull request merged**
